File: Topic03/016_queues/queues.js
class Queue {
  // private
  #queue = [];

  constructor() {}

  getLength = () => {
    return this.#queue.length;
  };

  isEmpty = () => {
    return this.#queue.length == 0 ? true : false;
  };

  head = () => {
    return !this.isEmpty()
      ? this.#queue[this.getLength() - 1]
      : "The queue is empty.";
  };

  enqueue = (num) => {
    this.#queue.unshift(num);
  };

  dequeue = () => {
    if (!this.isEmpty()) this.#queue.pop();
    else console.log("the queue is empty, nothing to remove.");
  };
}

/* example in video */
var que = new Queue();

// new born queue
console.log(`The queue's length : ${que.getLength()}`);

console.log(`The queue is empty: ${que.isEmpty()}`);

console.log(`The queue's head : ${que.head()}`);

// start adding elements into the queue

console.log("------\nAdding 1 into the queue :");

que.enqueue(1);

console.log(`The queue's length : ${que.getLength()}`);

console.log(`The queue is empty: ${que.isEmpty()}`);

console.log(`The queue's head : ${que.head()}`);

console.log("------\nAdding 2 into the queue :");

que.enqueue(2);

console.log(`The queue's length : ${que.getLength()}`);

console.log(`The queue is empty: ${que.isEmpty()}`);

console.log(`The queue's head : ${que.head()}`);

// remove one element

console.log("------\nRemoving one element into the queue :");

que.dequeue();

console.log(`The queue's length : ${que.getLength()}`);

console.log(`The queue is empty: ${que.isEmpty()}`);

console.log(`The queue's head : ${que.head()}`);

// ENQUEUE[1, q]

console.log("------\nAdding once more 1 into the queue :");

que.enqueue(1);

console.log(`The queue's length : ${que.getLength()}`);

console.log(`The queue is empty: ${que.isEmpty()}`);

console.log(`The queue's head : ${que.head()}`);

// x = HEAD[q]

var obj = { x: que.head() };

console.table(obj);

/*
/* review question 7 
*/

console.log("\n----------------------\n");
// new Queue q
var que2 = new Queue();

// ENQUEUE[1, q]
que2.enqueue(1);

console.log(`The queue's length : ${que2.getLength()}`);

console.log(`The queue is empty: ${que2.isEmpty()}`);

console.log(`The queue's head : ${que2.head()}`);

/* for 1 <= i <= 5 do
 *  ENQUEUE(i x HEAD[q])
 *  DEQUEUE[q]
 *  end for
 */

for (let i = 0; i < 5; i++) {
  que2.enqueue((i + 1) * que2.head());
  que2.dequeue();
}

console.log("------\nAdding elements into the queue :");

console.log(`The queue's length : ${que2.getLength()}`);

console.log(`The queue is empty: ${que2.isEmpty()}`);

console.log(`The queue's head : ${que2.head()}`);

// x = HEAD[q]

obj = { x: que2.head() };

console.table(obj);

/*
/* review question 9 
*/

console.log("\n----------------------\n");

/* function qFunc(n)
  for 1 <= i <= n do
    ENQUEUE[i mod 2, q]
  end for

  x = 0

  while EMPTY[q] = FALSE do
    x = x + HEAD[q]
    DEQUEUE[q]
  end while

  return x
  end function
 */

const qFunc = (n) => {
  let que = new Queue();

  for (let i = 0; i < n; i++) que.enqueue((i + 1) % 2);

  let x = 0;

  while (!que.isEmpty()) {
    x += que.head();
    que.dequeue();
  }

  return x;
};

// qFunc(5)
console.log(qFunc(5));

File: Data_Structures/016_queues/queues.py
class Queue:
    def __init__(self):
        # private
        self.__q = []

    def getLength(self):
        return len(self.__q)

    def head(self):
        if not(self.isEmpty()):
            return self.__q[-1]
        return "nothing in the queue"

    def isEmpty(self):
        return self.getLength() == 0

    def enqueue(self, num):
        self.__q.insert(0, num);

    def dequeue(self):
        if not(self.isEmpty()):
            self.__q.pop(-1)
        else:
            return "nothin in the queue"

# example in video

# new Queue que
que = Queue()

print("Setting up a empty queue : ")
print(f'Length of queue : {que.getLength()}')
print(f'the queue is empty : {que.isEmpty()}')
print(f'Head of queue : {que.head()}')

# ENQUEUE[1, q]
print("---\nAdd \'1\' into the queue : ")
que.enqueue(1)
print(f'Length of queue : {que.getLength()}')
print(f'the queue is empty : {que.isEmpty()}')
print(f'Head of queue : {que.head()}')

# ENQUEUE[2, q]
print("---\nAdd \'2\' into the queue : ")
que.enqueue(2)
print(f'Length of queue : {que.getLength()}')
print(f'the queue is empty : {que.isEmpty()}')
print(f'Head of queue : {que.head()}')

# DEQUEUE[q]
print("---\nRemove one item from the queue : ")
que.dequeue()
print(f'Length of queue : {que.getLength()}')
print(f'the queue is empty : {que.isEmpty()}')
print(f'Head of queue : {que.head()}')

# ENQUEUE[1, q]
print("---\nAdd \'1\' into the queue : ")
que.enqueue(1)
print(f'Length of queue : {que.getLength()}')
print(f'the queue is empty : {que.isEmpty()}')
print(f'Head of queue : {que.head()}')

# x = HEAD[q]
x = que.head()
print(f'---\nx : {x}')

# review question 7

'''
new Queue q
ENQUEUE[1, q]

for 1 <= i <= 5 do
  ENQUEUE(i x HEAD[q])
  DEQUEUE[q]
end for

x = HEAD[q]
'''
que2 = Queue()

que2.enqueue(1)

for i in range(5):
    que2.enqueue((i+1) * que2.head())
    que2.dequeue()

x = que2.head()
print(f'---\nx : {x}')

# review question 9

'''
function qFunc(n)
  for 1 <= i <= n do
    ENQUEUE[i mod 2, q]
  end for

  x = 0

  while EMPTY[q] = FALSE do
    x = x + HEAD[q]
    DEQUEUE[q]
  end while

  return x
end function
'''
def qFunc(n):
    que = Queue()
    
    for i in range(n):
        que.enqueue((i+1) % 2)

    x = 0

    while not que.isEmpty():
        x += que.head()
        que.dequeue();

    return x

# qFunc(5)

print(f'---\nqFunc(5) : {qFunc(5)}')
